README.md
# TAkka-Play 2.1

This project is forked from the [original Play repository](https://github.com/playframework/Play20) on 27 June 2013

In this project, Akka actors in the library is replaced by TAkka actors.

All changes in the TAKKA-PLAY branch are listed [here](https://github.com/Jiansen/Play20/blob/master/changes.md).


We follow the Play [Testing Method](http://www.playframework.com/documentation/2.1.1/BuildingFromSource) to validate the TAkka port.


# Addional Test
sample/scala/framework_benchmarks is modified from the [Play-Scala](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/play-scala) example in the [TechEmpower Web Framework Benchmarks](https://github.com/TechEmpower/FrameworkBenchmarks).

We tested the JSON benchmark using 

ab -kc 20 -t 60 http://localhost:9000/json

99% of 50000 requests finishes within 4ms using the Akka-Play or the TAkka-Play.





# Play 2.1 

Play 2.1 is a high productivity Java and Scala Web application framework, integrating all components and API needed for modern Web application development. 

It is based on a lightweight stateless Web friendly architecture and features predictable and minimal resources consumption (CPU, Memory, Threads) for highly scalable applications thanks to its reactive model based on Iteratee IO.

http://www.playframework.com

## Installing

You first need to build Play:

```bash
$ cd framework/
$ ./build
> publish-local
```

Then, for convenience, add the framework installation directory to your system PATH.  
On UNIX systems, it would be something like:

```bash
export PATH=$PATH:/path/to/play2.1
```

On windows systems you'll need to set it in the global environment variables.

> If you’re on UNIX, make sure that the play script is executable (otherwise do a chmod a+x play).

## Getting started

Enter any existing Play 2.1 application directory and use the `play` command to launch the development console:

```bash
$ cd ~/workspace/myApplication
$ play
```

You can also directly use `play run` to run the application:

```bash
$ cd ~/workspace/myApplication
$ play run
```

Use `play new yourNewApplication` to create a new application:

```bash
$ cd ~/workspace
$ play new myNewApplication
```

Once the application is created, use it as any existing application:

```bash
$ cd myNewApplication
$ play
```

## Running the sample applications

There are several samples applications included in the `samples/` directory. For example, to run the included Scala Hello World application:

```bash
$ cd ~/workspace/play2.1/samples/scala/helloworld/
$ play run
```
> The application will be available on port 9000.

## Issues tracker

Report issues at https://github.com/playframework/Play20/issues.

## Contributors

Check for all contributors at https://github.com/playframework/Play20/contributors.

## Licence

This software is licensed under the Apache 2 license, quoted below.

Copyright 2013 Typesafe (http://www.typesafe.com).

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this project except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


changes.md
Without deep understanding about the Play framework, I ported it to the TAkka version within less than 7 hours.  Following are changes I made to the Play source code.

# Changes to the Library Implementation

## play.core.j.OrderedExecutionContext.scala

line 3:
import akka.actor.{ Actor, ActorSystem, Props }  ->
import takka.actor.{ TypedActor, ActorSystem, Props }

line 17:
  private val actors = Array.fill(size)(actorSystem.actorOf(Props[OrderedExecutionContext.RunActor])) ->
  private val actors = Array.fill(size)(actorSystem.actorOf(Props[Runnable, OrderedExecutionContext.RunActor]))

line 33:
  class RunActor extends Actor { ->
  class RunActor extends TypedActor[Runnable] {




## play.core.j.OrderedExecutionContextSpec.scala

line 2: 
import akka.actor.ActorSystem  ->
import takka.actor.ActorSystem

line 34:
    def receive = {  ->
    def typedReceive = {

## play.core.Invoker.scala

line 3:
import akka.actor._ ->
import takka.actor._

## play.libs.Akka.java
### add play.libs.Akka.TAkka.java by changing
line 3:
import akka.actor.ActorSystem; ->
import takka.actor.ActorSystem; 

line 16:
public class Akka {  ->
public class TAkka {



line 32:
return play.api.libs.concurrent.Akka.system(Play.current());  ->
return play.api.libs.concurrent.TAkka.system(Play.current());
## play.api.libs.concurrent.Akka.scala
### add play.api.libs.concurrent.TAkka.scala by changing
line 6:
import akka.actor.ActorSystem ->
import takka.actor.ActorSystem

line 13:
object Akka { ->
object TAkka {

line 24:
app.plugin[AkkaPlugin].map(_.applicationSystem).getOrElse { ->
app.plugin[TAkkaPlugin].map(_.applicationSystem).getOrElse {

line 49
class AkkaPlugin(app: Application) extends Plugin { ->
class TAkkaPlugin(app: Application) extends Plugin {

## play.core.server.Server.scala

line 13 and 14:
import takka.actor._
import akka.actor.Actor._  ->

import takka.actor._
import takka.actor.Actor._


# sbt test passed

# ./runtest passed

## Add test for using TAKKA actors

### /documentation/manual/scalaGuide/main/akka/code/ScalaTAkka.scala

modified from /documentation/manual/scalaGuide/main/akka/code/ScalaAkka.scala

line 10:
import akka.actor.{Actor, Props} ->
import takka.actor.{TypedActor, Props}

line 14:
import play.api.libs.concurrent.Akka ->
import play.api.libs.concurrent.TAkka


line 28, 40, 52:
val testActor = Akka.system.actorOf(Props[MyActor], name = "testActor")  ->
val testActor = TAkka.system.actorOf(Props[String, MyActor], "testActor")

line 44:

Akka.system.scheduler.schedule(0.microsecond, 300.microsecond, testActor, "tick")  ->
TAkka.system.scheduler.schedule(0.microsecond, 300.microsecond, testActor, "tick")

line 58:
Akka.system.scheduler.scheduleOnce(1000.microsecond) { ->
TAkka.system.scheduler.scheduleOnce(1000.microsecond) {

line 72 - 76
class MyActor extends Actor {
  def typedReceive = {
    case s: String => sender ! "Hello," + s
  }
}

->

class MyActor extends TypedActor[String] {
  def typedReceive = {
    case s: String => sender ! "Hello," + s
  }
}



documentation/manual/scalaGuide/main/akka/code/ScalaTAkka.scala
package scalaguide.takka {


import org.specs2.mutable.Specification
import org.junit.runner.RunWith
import org.specs2.runner.JUnitRunner
import scala.concurrent.Await
import scala.concurrent.duration._

import takka.actor.{TypedActor, Props}
import akka.pattern.ask
import akka.util.Timeout

import play.api.libs.concurrent.TAkka
import play.api.Play.current
import play.api.test._
import play.api.test.Helpers._
import java.io.File

@RunWith(classOf[JUnitRunner])
class ScalaTAkkaSpec extends Specification {

  "A scala TAkka" should {
/*
    "myActor" in {
      running(FakeApplication()) {
        //#play-akka-myactor
        val myActor = TAkka.system.actorOf(Props[String, MyActor], "myactor")
        //#play-akka-myactor
        implicit val timeout = Timeout(300.microsecond)

        val future = myActor ? "Alan"
        val result = Await.result(future, 300.microsecond).asInstanceOf[String]
        result must contain("Hello")
      }
    }

    "actor scheduler" in {
      running(FakeApplication()) {
        val testActor = TAkka.system.actorOf(Props[String, MyActor], "testActor")
        import scala.concurrent.ExecutionContext.Implicits.global
        //#play-akka-actor-schedule-repeat
        import play.api.libs.concurrent.Execution.Implicits._
        TAkka.system.scheduler.schedule(0.microsecond, 300.microsecond, testActor, "tick")
        //#play-akka-actor-schedule-repeat
        1 === 1
      }
    }

    "actor scheduler" in {
      running(FakeApplication()) {
        val testActor = TAkka.system.actorOf(Props[String, MyActor], "testActor")
        import scala.concurrent.ExecutionContext.Implicits.global
        val file = new File("/tmp/nofile")
        file.mkdirs()
        //#play-akka-actor-schedule-run-once
        import play.api.libs.concurrent.Execution.Implicits._
        TAkka.system.scheduler.scheduleOnce(1000.microsecond) {
          file.delete()
        }
        //#play-akka-actor-schedule-run-once
        Thread.sleep(1005)
        file.exists() must beFalse
      }
    }
*/
  }


}

class MyActor extends TypedActor[String] {
  def typedReceive = {
    case s: String => sender ! "Hello," + s
  }
}

}
